shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -1681,11 +1687,36 @@ - (CGRect)firstRectForRange:(UITextRange*)range {
         !isLastSelectionRect && _selectionRects[i + 1].position > first;
     if (startsOnOrBeforeStartOfRange &&
         (endOfTextIsAfterStartOfRange || nextSelectionRectIsAfterStartOfRange)) {
-      return _selectionRects[i].rect;
+      // TODO(hellohaunlin): Remove iOS 17 check. The logic should also work for older versions.
+      if (@available(iOS 17, *)) {
+        startSelectionRect = _selectionRects[i].rect;
+      } else {
+        return _selectionRects[i].rect;
+      }
+    }
+    // TODO(hellohaunlin): Remove iOS 17 check. The logic should also work for older versions.
+    if (@available(iOS 17, *)) {
+      if (!CGRectIsNull(startSelectionRect)) {
+        BOOL endsOnOrAfterEndOfRange = _selectionRects[i].position >= end - 1;  // end is exclusive
+        BOOL nextSelectRectIsOnNextLine =
+            !isLastSelectionRect &&
+            _selectionRects[i + 1].rect.origin.y != _selectionRects[i].rect.origin.y;
```

Comment:
Is every character on the same line guaranteed to have the same `y` origin, or could this have false positives for unusual characters?